### PR TITLE
ceph: mgr dashboard service support externalIPs.

### PR DESCRIPTION
**Pull request by devlikai**
**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Review by leseb — CHANGES_REQUESTED** [review]
https://rook.io/docs/rook/master/development-flow.html#commit-structure

File: pkg/operator/ceph/cluster/mgr/spec.go [hunk 1]
```diff
@@ -223,6 +233,7 @@ func (c *Cluster) makeMgrSidecarContainer(mgrConfig *mgrConfig) v1.Container {
 			},
 		}},
 		{Name: "ROOK_DASHBOARD_ENABLED", Value: strconv.FormatBool(c.spec.Dashboard.Enabled)},
+		{Name: "ROOK_DASHBOARD_EXTERNAL_IPS", Value: externalIps},
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/spec.go:236]: With the `strings.Join()` we shouldn't need the for loop above
```suggestion
		{Name: "ROOK_DASHBOARD_EXTERNAL_IPS", Value: strings.Join(externalIps, ",")},
```

File: pkg/operator/ceph/cluster/mgr/mgr.go [hunk 2]
```diff
@@ -243,6 +244,13 @@ func (c *Cluster) ReconcileActiveMgrServices(daemonNameToUpdate string) error {
 	} else {
 		currentDaemon := svc.Spec.Selector[controller.DaemonIDLabel]
 		if currentDaemon == daemonNameToUpdate {
+			if corev1.ServiceTypeClusterIP == svc.Spec.Type {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/mgr/mgr.go:247]: Does this code block have any side effect? If it's just logging, seems like we don't need it?

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 3]
```diff
@@ -217,6 +217,11 @@ type ClusterSpec struct {
 	// +optional
 	// +nullable
 	LogCollector LogCollectorSpec `json:"logCollector,omitempty"`
+
+	// specify externalIPs
+	// +optional
+	// +nullalbe
+	ExternalIps []string `json:"externalIPs,omitempty"`
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:224]: Let's add this in the `DashboardSpec` instead of the `ClusterSpec`

File: cmd/rook/ceph/mgr.go [hunk 4]
```diff
@@ -67,6 +69,15 @@ func init() {
 func runMgrSidecar(cmd *cobra.Command, args []string) error {
 	rook.SetLogLevel()
 
+	logger.Infof("runMgrSidecar get ExternalIps:%s", dashboardExternalIps)
+	if dashboardExternalIps != "" {
+		for _, ip := range strings.Split(dashboardExternalIps, ",") {
+			if ip != "" {
+				clusterSpec.Dashboard.ExternalIps = append(clusterSpec.Dashboard.ExternalIps, ip)
```
Comment (travisn) [inline cmd/rook/ceph/mgr.go:76]: You could just assign the results of `Split()` directly and no need for the loop, right?
```suggestion
		clusterSpec.Dashboard.ExternalIps = strings.Split(clusterSpec.Dashboard.ExternalIps, ",")
```

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 3 above)
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:224]: You'll need to run `make codegen` and `make crds` to generate the artifacts needed for this setting.

**Review by travisn — CHANGES_REQUESTED** [review]

**Comment by leseb** [comment]
@devlikai any plan to resume this?

**Comment by BlaineEXE** [comment]
The commits should have an explanatory body message. Ideally, this would explain the scenario that is enabled by these changes and how it allows the scenario. Commits should probably be squashed together and have a signed-off-by line. Also, do not end a commit subject in punctuation.

**Pull request closed without merging**